plot/plot.go

```diff
@@ -0,0 +1,314 @@
+		Kind:   kind,
+		Name:   name,
+		Title:  name,
+		XLabel: "Value",
+	}
+}
+
+func (g *Graph) SetTitle(t string) *Graph {
+	g.Title = t
+	return g
+}
+
+func (g *Graph) SetXLabel(l string) *Graph {
+	g.XLabel = l
+	return g
+}
+
+func (g *Graph) SetYLogScale(b bool) *Graph {
+	g.YLogScale = b
+	return g
+}
+
+// AddPlotRight adds a plot to be displayed using the right Y axis. It's an
+// error if the plot and the Graph have different Kinds.
+func (g *Graph) AddPlotRight(p *Plot) error {
+	if p.Kind != g.Kind {
+		return errors.Reason("plot's kind [%s] != graph's kind [%s]",
+			p.Kind, g.Kind)
+	}
+	g.PlotsRight = append(g.PlotsRight, p)
+	return nil
+}
+
+// AddPlotLeft adds a plot to be displayed using the left Y axis. It's an error
+// if the plot and the Graph have different Kinds.
+func (g *Graph) AddPlotLeft(p *Plot) error {
+	if p.Kind != g.Kind {
+		return errors.Reason("plot's kind [%s] != graph's kind [%s]",
+			p.Kind, g.Kind)
+	}
+	g.PlotsLeft = append(g.PlotsLeft, p)
+	return nil
+}
+
+// Group of Graphs sharing the same X axis. Visually, all graphs should be
+// displayed one after another in the given order, having the same width and
+// aligned vertically, to match their X axis positions.
+type Group struct {
+	Kind      Kind              // must match Graph's Kind
+	Name      string            // internal unique name of the group
+	XLogScale bool              // whether to use log-scale for X axis
+	Graphs    []*Graph          // to preserve the order
+	graphMap  map[string]*Graph // for quick access
+}
+
+func NewGroup(kind Kind, name string) *Group {
+	return &Group{
+		Kind:     kind,
+		Name:     name,
+		graphMap: make(map[string]*Graph),
+	}
+}
+
+func (g *Group) SetXLogScale(b bool) *Group {
+	g.XLogScale = b
+	return g
+}
+
+// addGraph to both the slice and the map, to keep them in sync.
+func (g *Group) addGraph(graph *Graph) {
+	g.Graphs = append(g.Graphs, graph)
+	g.graphMap[graph.Name] = graph
+}
+
+// AddGraph to the Group. It's an error if the Graph Kind doesn't match the
+// Group Kind.
+func (g *Group) AddGraph(graph *Graph) error {
+	if g.Kind != graph.Kind {
+		return errors.Reason("group's Kind [%s] != graph's Kind [%s]",
+			g.Kind, graph.Kind)
+	}
+	if _, ok := g.graphMap[graph.Name]; ok {
+		return errors.Reason("graph %s already exists in group %s",
+			graph.Name, g.Name)
+	}
+	g.addGraph(graph)
+	return nil
+}
+
+// Canvas is the master collection for all the plot groups
+type Canvas struct {
+	Groups   []*Group          // to preserve the order
+	groupMap map[string]*Group // for quick reference by name
+	graphMap map[string]*Graph // direct Graph reference within Groups
+}
+
+func NewCanvas() *Canvas {
+	return &Canvas{
+		groupMap: make(map[string]*Group),
+		graphMap: make(map[string]*Graph),
+	}
+}
+
+// addGroup to both the slice and the map, to keep them in sync.
+func (c *Canvas) addGroup(group *Group) {
+	c.Groups = append(c.Groups, group)
+	c.groupMap[group.Name] = group
+}
+
```

Comment:
Add `GetGroup` and `GetGraph` that would return `nil` if not present. These can be used by the caller to ensure their presence, and add if necessary.

In fact, `EnsureGraph` can do exactly that: create and add an instance as necessary both for the group and graph, and return the `Graph` instance for further configuration.